src/audio/copier.c

```diff
@@ -0,0 +1,615 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2021 Intel Corporation. All rights reserved.
+//
+// Author: Rander Wang <rander.wang@linux.intel.com>
+
+#include <sof/audio/buffer.h>
+#include <sof/audio/component_ext.h>
+#include <sof/audio/format.h>
+#include <sof/audio/pipeline.h>
+#include <sof/common.h>
+#include <sof/debug/panic.h>
+#include <sof/drivers/interrupt.h>
+#include <sof/ipc/msg.h>
+#include <sof/ipc/topology.h>
+#include <sof/drivers/interrupt.h>
+#include <sof/drivers/timer.h>
+#include <sof/lib/alloc.h>
+#include <sof/lib/cache.h>
+#include <sof/lib/memory.h>
+#include <sof/lib/notifier.h>
+#include <sof/lib/uuid.h>
+#include <sof/list.h>
+#include <sof/string.h>
+#include <sof/ut.h>
+#include <sof/trace/trace.h>
+#include <ipc4/alh.h>
+#include <ipc4/copier.h>
+#include <ipc4/module.h>
+#include <ipc4/error_status.h>
+#include <ipc4/gateway.h>
+#include <ipc/dai.h>
+#include <user/trace.h>
+#include <errno.h>
+#include <stddef.h>
+#include <stdint.h>
+
+static const struct comp_driver comp_copier;
+
+/* this id aligns windows driver requirement to support windows driver */
+/* 9ba00c83-ca12-4a83-943c-1fa2e82f9dda */
+DECLARE_SOF_RT_UUID("copier", copier_comp_uuid, 0x9ba00c83, 0xca12, 0x4a83,
+		    0x94, 0x3c, 0x1f, 0xa2, 0xe8, 0x2f, 0x9d, 0xda);
+
+DECLARE_TR_CTX(copier_comp_tr, SOF_UUID(copier_comp_uuid), LOG_LEVEL_INFO);
+
+struct copier_data {
+	struct ipc4_copier_module_cfg config;
+	struct comp_dev *endpoint;
+	int direction;
+
+	struct ipc4_audio_format out_fmt[IPC4_COPIER_MODULE_OUTPUT_PINS_COUNT];
+	pcm_converter_func converter[IPC4_COPIER_MODULE_OUTPUT_PINS_COUNT];
+};
+
+static enum sof_ipc_frame convert_fmt(uint32_t depth, uint32_t type)
+{
+	switch (depth) {
+	case IPC4_DEPTH_16BIT:
+		return SOF_IPC_FRAME_S16_LE;
+	case IPC4_DEPTH_24BIT:
+		return SOF_IPC_FRAME_S24_4LE;
+	case IPC4_DEPTH_32BIT:
+		if (type == IPC4_TYPE_FLOAT)
```

Comment:
feel free to completely ignore cause this is a massive nit pick, but a ternary could also clean this up

Reply:
@RanderWang this can be fixed in an incremental PR as further feature integration will change this code.